buzhash.go

```diff
@@ -0,0 +1,115 @@
+package chunk
+
+import (
+	"io"
+	"math/bits"
+
+	pool "github.com/libp2p/go-buffer-pool"
+)
+
+const (
+	buzMin  = 128 << 10
+	buzMax  = 512 << 10
+	buzMask = 1<<17 - 1
+)
+
+type Buzhash struct {
+	r   io.Reader
+	buf []byte
+	n   int
+
+	err error
+}
+
+func NewBuzhash(r io.Reader) *Buzhash {
+	return &Buzhash{
+		r:   r,
+		buf: pool.Get(buzMax),
+	}
+}
+
+func (b *Buzhash) NextBytes() ([]byte, error) {
+	if b.err != nil {
+		return nil, b.err
+	}
+
+	buf := b.buf
+	n, err := io.ReadFull(b.r, buf[b.n:])
+	if err != nil {
+		if err == io.ErrUnexpectedEOF {
+			b.err = io.EOF
+			return buf[:n+b.n], nil
```

Comment:
If the block is smaller than _buzMax_ we return it immediately? Is that right?

Reply:
Yes, it is a simplification that allows it to be faster and simpler.

Reply:
It can't be that much more complicated to return if <= buzMin. and continue otherwise. Are you doing this for loop unrolling?

Reply:
I thought it will be more complex to handle. Running benchmarks now.

Reply:
Overhead is 9%:
```
name       old time/op    new time/op    delta
Buzhash-4    14.2ms ± 7%    15.6ms ± 8%   +9.83%  (p=0.000 n=20+18)

name       old speed      new speed      delta
Buzhash-4  1.18GB/s ± 6%  1.07GB/s ±10%   -9.45%  (p=0.000 n=20+19)
```

Reply:
Can you post the code? Is this 9% on top of changing the constants?

Reply:
Yes, it is another 9%

Reply:
I was able to eliminate it the perf hit by adding a conditional panic that never fires (it probably reassures the compiler).

```
name             old time/op    new time/op    delta
Buzhash2/1K-4       592ns ± 1%     594ns ± 5%     ~     (p=0.920 n=9+10)
Buzhash2/1M-4       911µs ± 3%    1261µs ± 2%  +38.48%  (p=0.000 n=9+9)
Buzhash2/16M-4     18.8ms ± 4%    18.6ms ± 4%     ~     (p=0.190 n=10+10)
Buzhash2/100M-4     117ms ± 1%     114ms ± 2%   -2.77%  (p=0.000 n=10+9)

name             old speed      new speed      delta
Buzhash2/1K-4    1.73GB/s ± 1%  1.72GB/s ± 5%     ~     (p=0.888 n=9+10)
Buzhash2/1M-4    1.15GB/s ± 3%  0.83GB/s ± 2%  -27.80%  (p=0.000 n=9+9)
Buzhash2/16M-4    891MB/s ± 4%   901MB/s ± 4%     ~     (p=0.190 n=10+10)
Buzhash2/100M-4   894MB/s ± 1%   919MB/s ± 2%   +2.85%  (p=0.000 n=10+9)

name             old alloc/op   new alloc/op   delta
Buzhash2/1K-4      1.17kB ± 0%    1.17kB ± 0%     ~     (p=0.088 n=10+10)
Buzhash2/1M-4      1.09MB ± 1%    1.08MB ± 0%   -0.90%  (p=0.000 n=10+8)
Buzhash2/16M-4     17.1MB ± 0%    17.1MB ± 0%     ~     (p=0.971 n=10+10)
Buzhash2/100M-4     106MB ± 0%     106MB ± 0%   +0.00%  (p=0.006 n=10+9)

name             old allocs/op  new allocs/op  delta
Buzhash2/1K-4        3.00 ± 0%      3.00 ± 0%     ~     (all equal)
Buzhash2/1M-4        7.00 ± 0%      8.00 ± 0%  +14.29%  (p=0.000 n=10+10)
Buzhash2/16M-4       71.0 ± 0%      72.0 ± 0%   +1.41%  (p=0.000 n=10+10)
Buzhash2/100M-4       405 ± 0%       406 ± 0%     ~     (p=0.370 n=10+10)
```

Reply:
And it is in the PR.